---
title: Lỗi
slug: Web/JavaScript/Reference/Global_Objects/loi
tags:
  - JavaScript
  - Lỗi
  - Tham chiếu
translation_of: Web/JavaScript/Reference/Global_Objects/Error
---
<div>{{JSRef}}</div>

<div>Hàm <strong><code>Error</code></strong> tạo ra một đối tượng kiểu lỗi. Các thể hiện của đối tượng <code>Error</code> được ném ra khi xảy ra lỗi thực thi. Một đối tượng <code>Error</code> ngoài ra có thể được dùng như là một đối tượng căn bản cho các ngoại lệ do người dùng tự định nghĩa. Xem bên dưới các loại lỗi chuẩn được xây dựng sẵn.</div>

<h2 id="Cú_pháp">Cú pháp</h2>

<pre class="syntaxbox">new Error([<var>message</var>[, <var>fileName</var>[, <var>lineNumber</var>]]])</pre>

<h3 id="Tham_số">Tham số</h3>

<dl>
 <dt><code>message</code></dt>
 <dd>Tùy chọn. Một mô tả của lỗi rằng người dùng có thể đọc được.</dd>
 <dt><code>fileName</code> {{non-standard_inline}}</dt>
 <dd>Tùy chọn. Giá trị của thuộc tính <code>fileName</code> trong quá trình tạo ra đối tượng <code>Error</code>. Các mặc định là tên của tập tin chứa đoạn mã gọi tới hàm <code>Error()</code>.  </dd>
 <dt><code>lineNumber</code> {{non-standard_inline}}</dt>
 <dd>Tuỳ chọn. Giá trị của thuộc tính <code>lineNumber</code> trong quá trình tạo ra đối tượng <code>Error</code>. Các mặc định là số dòng hiện tại đang gọi tới hàm <code>Error()</code>.</dd>
</dl>

<h2 id="Mô_tả">Mô tả</h2>

<p>Các lỗi thực thi dẫn đến các đối tượng <code>Error</code> được tạo ra và ném đi.</p>

<p>Trang này ghi lại việc sử dụng đối tượng Error và cách sử dụng nó như là một hàm. Để biết danh sách các thuộc tính và phương thức được kế thừa bởi các thể hiện của <code>Error</code>, xem {{jsxref("Error.prototype")}}.</p>

<h3 id="Được_sử_dụng_như_là_một_hàm">Được sử dụng như là một hàm</h3>

<p>Khi <code>Error</code> được sử dụng như là một hàm -- không có từ khoá <code>new</code>, nó sẽ trả về một đối tượng <code>Error</code>. Do đó, gọi hàm <code>Error</code> cũng sẽ tạo ra cùng một đối tượng <code>Error</code> thông qua từ khóa <code>new</code>.</p>

<pre>// this:
const x = Error('Tôi được tạo ra bằng cách sử dụng gọi hàm!');<code>
​​​​// tương tự như this</code>
<code>const y = new Error('Tôi được xây dựng thông qua từ khoá "new"!');</code></pre>

<h3 id="Các_loại_lỗi">Các loại lỗi</h3>

<p>Bên cạnh hàm Error, còn có 7 hàm khác làm việc với lỗi trong JavaScript. Đối với ngoại lệ phía client, xem thêm tại <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Statements#Exception_handling_statements">Các câu lệnh xử lý ngoại lệ.</a></p>

<dl>
 <dt>{{jsxref("EvalError")}}</dt>
 <dd>Tạo một thể hiện đại diện cho một lỗi xảy ra liên quan đến hàm toàn cục {{jsxref("Global_Objects/eval", "eval()")}}.</dd>
 <dt>{{jsxref("InternalError")}} {{non-standard_inline}}</dt>
 <dd>Tạo một thể hiện đại diện cho một lỗi xảy ra khi một lỗi bên trong bộ máy JavaScript được ném. Ví dụ: "quá nhiều đệ quy".</dd>
 <dt>{{jsxref("RangeError")}}</dt>
 <dd>Tạo một thể hiện đại diện cho một lỗi xảy ra khi một biến số hoặc tham số nằm ngoài phạm vi hợp lệ của nó.</dd>
 <dt>{{jsxref("ReferenceError")}}</dt>
 <dd>Tạo một thể hiện đại diện cho một lỗi xảy ra khi hủy tham chiếu của một tham chiếu không hợp lệ.</dd>
 <dt>{{jsxref("SyntaxError")}}</dt>
 <dd>Tạo một thể hiện đại diện cho một lỗi xảy ra trong khi phân tích cú pháp mã trong {{jsxref("Global_Objects/eval", "eval()")}}.</dd>
 <dt>{{jsxref("TypeError")}}</dt>
 <dd>Tạo một thể hiện đại diện cho một lỗi xảy ra khi một biến hoặc một tham số có kiểu không hợp lệ.</dd>
 <dt>{{jsxref("URIError")}}</dt>
 <dd>Tạo một thể hiện đại diện cho một lỗi xảy ra khi {{jsxref("Global_Objects/encodeURI", "encodeURI()")}} hoặc {{jsxref("Global_Objects/decodeURI", "decodeURI()")}} truyền các tham số không hợp lệ.</dd>
</dl>

<h2 id="Thuộc_tính">Thuộc tính</h2>

<dl>
 <dt>{{jsxref("Error.prototype")}}</dt>
 <dd>Cho phép thêm các thuộc tính mới vào các thể hiện của <code>Error</code>.</dd>
</dl>

<h2 id="Phương_thức">Phương thức</h2>

<p>Một đối tượng <code>Error</code> toàn cục không chứa phương thức riêng của nó, tuy nhiên, nó thừa hưởng các phương thức từ prototype.</p>

<h2 id="Các_thể_hiện_của_Error">Các thể hiện của Error</h2>

<div>{{page('en-US/docs/JavaScript/Reference/Global_Objects/Error/prototype', 'Description')}}</div>

<h3 id="Các_thuộc_tính">Các thuộc tính</h3>

<div>{{page('en-US/docs/JavaScript/Reference/Global_Objects/Error/prototype', 'Properties')}}</div>

<h3 id="Các_phương_thức">Các phương thức</h3>

<div>{{page('en-US/docs/JavaScript/Reference/Global_Objects/Error/prototype', 'Methods')}}</div>

<h2 id="Các_ví_dụ">Các ví dụ</h2>

<h3 id="Ném_một_lỗi_thông_thường">Ném một lỗi thông thường</h3>

<p>Thông thường bạn tạo ra một đối tượng <code>Error</code> và muốn sử dụng nó cùng tới từ khoá {{jsxref("Statements/throw", "throw")}}. Bạn có thể bẫy lỗi bằng cách sử dụng cú pháp {{jsxref("Statements/try...catch", "try...catch")}}:</p>

<pre class="brush: js">try {
  throw new Error('Whoops!');
} catch (e) {
  console.log(e.name + ': ' + e.message);
}
</pre>

<h3 id="Xử_lý_một_lỗi_cụ_thể">Xử lý một lỗi cụ thể</h3>

<p>Bạn có thể chọn cách chỉ xử lý các loại lỗi cụ thể bằng cách kiểm tra loại lỗi với thuộc tính {{jsxref("Object.prototype.constructor", "constructor")}} của lỗi hoặc, nếu bạn đang viết cho các công cụ JavaScript hiện đại, hãy dùng từ khoá {{jsxref("Operators/instanceof", "instanceof")}}:</p>

<pre class="brush: js">try {
  foo.bar();
} catch (e) {
  if (e instanceof EvalError) {
    console.log(e.name + ': ' + e.message);
  } else if (e instanceof RangeError) {
    console.log(e.name + ': ' + e.message);
  }
  // ... etc
}
</pre>

<h3 id="Tùy_biến_các_loại_lỗi">Tùy biến các loại lỗi</h3>

<p>Bạn có thể muốn định nghĩa các loại lỗi của riêng mình được dẫn xuất từ <code>Error</code> để có thể <code>throw new MyError()</code> và sử dụng <code>instanceof MyError</code> để kiểm tra loại lỗi trong trình xử lý ngoại lệ. Điều này làm cho mã xử lý lỗi sạch hơn và nhất quán hơn. Xem "<a href="http://stackoverflow.com/questions/1382107/whats-a-good-way-to-extend-error-in-javascript">Cách tốt để thừa kế hàm Error trong JavaScript là gì?</a>" trên StackOverflow để thảo luận chuyên sâu.</p>

<h4 id="Tuỳ_biến_lớp_Error_trong_ES6">Tuỳ biến lớp Error trong ES6</h4>

<div class="warning">
<p>Babel và các trình biên dịch khác sẽ không xử lý chính xác đoạn mã sau mà không cần <a href="https://github.com/loganfsmyth/babel-plugin-transform-builtin-extend">cấu hình bổ sung</a><a href="/vi/docs/">.</a></p>
</div>

<div class="note">
<p>Một số trình duyệt đã thêm hàm CustomError trong ngăn xếp qua cách dùng class trong ES6.</p>
</div>

<pre class="brush: js">class CustomError extends Error {
  constructor(foo = 'bar', ...params) {
    // Pass remaining arguments (including vendor specific ones) to parent constructor
    super(...params);

    // Maintains proper stack trace for where our error was thrown (only available on V8)
    if (Error.captureStackTrace) {
      Error.captureStackTrace(this, CustomError);
    }

    // Custom debugging information
    this.foo = foo;
    this.date = new Date();
  }
}

try {
  throw new CustomError('baz', 'bazMessage');
} catch(e){
  console.log(e.foo); //baz
  console.log(e.message); //bazMessage
  console.log(e.stack); //stacktrace
}</pre>

<h4 id="Tuỳ_biến_đối_tượng_trong_ES5">Tuỳ biến đối tượng trong ES5</h4>

<div class="warning">
<p><strong>Tất cả</strong> trình duyệt đã thêm hàm CustomError trong ngăn xếp qua cách khai báo chuẩn.</p>
</div>

<pre class="brush: js">function CustomError(foo, message, fileName, lineNumber) {
  var instance = new Error(message, fileName, lineNumber);
  instance.foo = foo;
  Object.setPrototypeOf(instance, Object.getPrototypeOf(this));
  if (Error.captureStackTrace) {
    Error.captureStackTrace(instance, CustomError);
  }
  return instance;
}

CustomError.prototype = Object.create(Error.prototype, {
  constructor: {
    value: Error,
    enumerable: false,
    writable: true,
    configurable: true
  }
});

if (Object.setPrototypeOf){
  Object.setPrototypeOf(CustomError, Error);
} else {
  CustomError.__proto__ = Error;
}


try {
  throw new CustomError('baz', 'bazMessage');
} catch(e){
  console.log(e.foo); //baz
  console.log(e.message) ;//bazMessage
}</pre>

<h2 id="Các_đặc_điểm_kỹ_thuật">Các đặc điểm kỹ thuật</h2>

<table class="standard-table">
 <tbody>
  <tr>
   <th scope="col">Đặc điểm kỹ thuật</th>
   <th scope="col">Trạng thái</th>
   <th scope="col">Chú thích</th>
  </tr>
  <tr>
   <td>{{SpecName('ES1')}}</td>
   <td>{{Spec2('ES1')}}</td>
   <td>Định nghĩa ban đầu. Được triển khai trong JavaScript 1.1.</td>
  </tr>
  <tr>
   <td>{{SpecName('ES5.1', '#sec-15.11', 'Error')}}</td>
   <td>{{Spec2('ES5.1')}}</td>
   <td> </td>
  </tr>
  <tr>
   <td>{{SpecName('ES6', '#sec-error-objects', 'Error')}}</td>
   <td>{{Spec2('ES6')}}</td>
   <td> </td>
  </tr>
  <tr>
   <td>{{SpecName('ESDraft', '#sec-error-objects', 'Error')}}</td>
   <td>{{Spec2('ESDraft')}}</td>
   <td> </td>
  </tr>
 </tbody>
</table>

<h2 id="Tính_tương_thích_của_trình_duyệt">Tính tương thích của trình duyệt</h2>

<div>
<div class="hidden">Bảng tương thích trên trang này được tạo từ dữ liệu có cấu trúc. Nếu bạn muốn đóng góp cho dữ liệu, vui lòng xem <a href="https://github.com/mdn/browser-compat-data">https://github.com/mdn/browser-compat-data</a> và gửi cho yêu cầu cho chúng tôi.</div>

<p>{{Compat("javascript.builtins.Error")}}</p>
</div>

<h2 id="Tham_khảo">Tham khảo</h2>

<ul>
 <li>{{jsxref("Error.prototype")}}</li>
 <li>{{jsxref("Statements/throw", "throw")}}</li>
 <li>{{jsxref("Statements/try...catch", "try...catch")}}</li>
</ul>
